### ISSUE TITLE
Improvement: Support for frameworks which rely on ClassLoaders in AOT native compilation?

### ISSUE THREAD

**Issue by InfoSec812** [graalvm/graal-core#276]
I was wondering if it is possible or feasible to analyze a running application so that ClassLoaders have a chance to resolve their requirements at runtime and THEN use the profiling information to determine the type graph for native compilation? It wouldn't help IMMEDIATELY for native startup times, but it would make AOT/Graal much easier to consume.

**Comment by dougxc** [comment]
I'm not sure I understand what you are proposing. What ClassLoader requirements are you thinking of? And how would analysis help? The Graal compiler already using the class hierarchy information provided by the VM when compiling.

**Comment by InfoSec812** [comment]
So, for example, In a [Vert.x](http://vertx.io) application, they use a `Launcher` class which gets a parameter telling it which class will be used as the "Main Verticle". Without passing in the name of the Main Verticle, the Launcher class and the associated fat-jar cannot be compiled to a native executable using Graal. If, as a possible option, it would be possible to run the application and capture the classloader work which happens during the application's run; it might be possible to resolve the classloader loading and then generate the native executable.

**Comment by InfoSec812** [comment]
And I will admit that my understanding of JVM internals is limited.

**Comment by dougxc** [comment]
The latest GraalVM release includes a `native-image` command for producing a native image via AOT compilation. Extensive documentation for use of this tool is being prepared but I cannot yet give a date for when it will be publicly available. @christianwimmer might be able to say more about whether the tool can be configured to make certain parameters constant based on profiles of past executions.

**Comment by InfoSec812** [comment]
Yes, I am sorry for not being more clear. I was definitely referring to the `native-image` command. Right now, it takes a parameter of the `Main` class which has a `public static void main()` method. Vert.x's `Launcher` class has this method, but it requires certain arguments in order to load the required classes. Thus, it would not work correctly with `native-image`.

**Comment by smarr** [comment]
The `static {}` blocks of classes are executed during image generation. If you could ensure that Vert.x is properly configured at that time, and did load the classes you need, I think, you might get the desired effect.

**Comment by InfoSec812** [comment]
@smarr I could `extend` `io.vertx.core.Launcher` in order to get it to load my classes, but for the more general case it would be more user-friendly if there were a more generic way to handle runtime classloading.
